### ISSUE TITLE
Help us with better command and parameter/flag descriptions

### ISSUE THREAD

**Issue by fdncred** [nushell/nushell#5066]
### Related problem

Some of our descriptions of commands and parameter descriptions could be more informative. Now that we've added descriptions in our completions, it would be nice to have them be as informative as possible without being too verbose.

Example of command descriptions
<img width="950" alt="Screen Shot 2022-04-01 at 4 12 46 PM" src="https://user-images.githubusercontent.com/343840/161342205-8c86c8e5-2e46-4292-a53e-2304e52f6a70.png">

Example of parameter/flag descriptions
<img width="837" alt="Screen Shot 2022-04-01 at 4 13 18 PM" src="https://user-images.githubusercontent.com/343840/161342256-dce05397-4efd-439a-81b4-45adcd06ea2e.png">

### Describe the solution you'd like

We'd like to have helpful, concise descriptions for parameters/flags, and commands.

**Comment by wmstack** [comment]
I would love it if `str append` was added. Intuitively this would be the method for one to add a string to another string, rather than the infix `+` which is awkward to use in a pipeliny fashion.

Of course, one solution would be to introduce something like `|>` akin to the one being proposed for js:

```nu
echo "hello" |> ? + " world!"
```

**Comment by hyiltiz** [comment]
As a minor suggestion: all descriptions should follow the same grammar rule. E.g. `trims text` starts with lowercase letter `t` while `Display this help message` starts with uppercase `D`. For non-descriptions such as `deprecated command`, either put them in special symbols like `(deprecated command)`, or better, only hide them from the completion list but warn users who still use them. This way, new users are not incentivised or taught to use deprecated commands.

**Comment by fdncred** [comment]
@hyiltiz consistency would be a nice add.

**Comment by yhm-amber** [comment]
For most of gnu/linux command, they have the `--` flag, like this:

![S21208-16031405](https://user-images.githubusercontent.com/103625580/206392161-ab5da3f4-2a41-411e-b151-d7c93a64e358.png)

It divide the parameters and flags in args.

this

~~~ sh
cat -n foo path/to/bar baz
~~~

will be parsed as

~~~ sh
cat -n -- foo path/to/bar baz
~~~

and if you write `cat -n -- foo path/to/bar baz` in your code, it csn be more clearly: the `foo` is not a value for `-n` flag.

Here, `cat` command is simple enough, so even users see code like `cat -n foo` , they can just run `cat --help` and spend time to read it and they will understand `foo` is not a value for `-n` but a param.

But, when some command have flag that can give zero or one value to it, if don't use `--` , confuse will be easy.

So, I think make `ls -a -- .` and all the other command in Nushell support `--` flag/option, could be a good idea. 🙃

**Comment by jaudiger** [comment]
@fdncred I was looking into the outputs of some help commands. And I noticed that sometimes the explanation of a command (aka `usage`) finishes by a dot, and sometimes it doesn't:

![image](https://user-images.githubusercontent.com/31616285/217375838-0ed4972d-649f-46ee-81bd-e87e917a1e25.png)

![image](https://user-images.githubusercontent.com/31616285/217375952-a2720093-40fa-4c3c-8543-0e429737930e.png)

Same thing applies for `extra_usage`.

Also, there is an inconsistency with the arguments' description of some commands (below the command http get), some descriptions begin by a capital, where others don't:

![image](https://user-images.githubusercontent.com/31616285/217376831-0398887e-69e4-4303-9389-45b757d35e56.png)

I compared those outputs with ones from `curl` and `wget`, and they finish their sentence with a dot (which seems completely fine).

![image](https://user-images.githubusercontent.com/31616285/217377115-1828ecaf-20b3-4ce5-b226-779de1f4581e.png)

So, I would recommend the following things:

- `usage`: begin by a capital and end with a dot
- `extra_usage`: begin by a capital and end with a dot
- Argument `desc`: begin by a capital and do not end with a dot

Now, the last remaining thing: should we apply the same rules to the description of the examples? Should we begin those descriptions by a capital and prevent the ending with a dot?

**Comment by fdncred** [comment]
@jaudiger Agreed. Some consistency would be nice. I think the example descriptions should have some uniformity too.

**Comment by amtoine** [comment]
with my new interest in `help` and as a continuation to #8189, i wanted to go through the commands and look at their help and try to improve their messages and their examples!

**is there some work being done here?**
if not, i'd like to help making the help a bit better as i can :relieved:

**Comment by jaudiger** [comment]
@amtoine  I just opened an MR to begin to work on the first two bullets points cited in my last comment. As for now, I just update the messages to make sure they all end with the same terminal character.

**Comment by amtoine** [comment]
> @amtoine I just opened an MR to begin to work on the first two bullets points cited in my last comment. As for now, I just update the messages to make sure they all end with the same terminal character.
>
nice, that's cool thanks for the pointer :yum:

**Comment by jntrnr** [comment]
@fdncred - we may want to turn this into a list of commands that need better help text so people can find where to jump in a bit more easily.

**Comment by fdncred** [comment]
@jntrnr maybe once we have a 1.0 list of commands. even still, having a list of every command and all their parameters may be unwieldy.

**Comment by Yakiyo** [comment]
i think help messages should all be aligned. It makes its more presentable and easier on the eyes. From jaudiger's comment above (https://github.com/nushell/nushell/issues/5066#issuecomment-1421528910) the help message in wget looks much much better than the ones from nu's help messages.

**Comment by amtoine** [comment]
@Yakiyo and @jaudiger

i thought this was a good idea and started https://github.com/nushell/nushell/pull/10065 :yum:

**Comment by drbrain** [comment]
#11285 adds enforcement of positional arguments starting with an uppercase character and ending with a period.

While making the tests pass I noticed that most commands have positional argument descriptions that are minimally descriptive like "The thing that does stuff" a smaller amount have "Thing that does the stuff" while a few have single-word descriptions for like "Thing".

A good example of the difference is [`export def`'s required arguments]( https://github.com/nushell/nushell/blob/d0b73f4d4dfad3b83f35299010b7c361a6a71e9a/crates/nu-cmd-lang/src/core_commands/export_def.rs#L22-L24):

```rust
.required("def_name", SyntaxShape::String, "Command name.")
.required("params", SyntaxShape::Signature, "Parameters.")
.required("block", SyntaxShape::Block, "Body of the definition.")
```

Single-word definitions that repeat the argument name aren't very helpful. For consistency these should be slightly more wordy like:

* "The command name to define"
* "The command parameters"
* "The body of the command"

But maybe omitting "The"/"A"/"An" before providing a description for the argument is OK?

**Comment by amtoine** [comment]
@drbrain 
in the case of things like `export def`, i'm not even sure the help is useful at all tbh :laughing: 
the slightly bigger versions you propose definitely sound better than what we have now!

maybe we could explain a bit more?
- _The command parameters, a comma-separated list inside `[]`_
- _The body of the command, a list of instructions inside `{}`_

**Comment by xpe** [comment]
@drbrain I'm wondering if it would be good to close this out?

**Comment by kantnero** [comment]
Hi @fdncred,
I noticed this issue is labeled “good first issue” but it’s quite old. I’d like to take a look at it if it’s still relevant.

**Comment by fdncred** [comment]
hi @kantnero, you're welcome to go through the commands and parameters to make sure they're well worded, punctuated correctly, consistently have (or don't have) ending periods, fill in things that are missing, etc.

**Comment by monigarr** [comment]
I’m actively working on this and should have a PR up within a few hours tonight (January 20th 2026).

Scope is strictly limited to improving help/completion text only (string literals): making command, argument, flag, and example descriptions more informative and consistent, with no logic or structural changes. 

I’m working file-by-file and running existing tests to keep this low-risk.   Tests are still running as of 5:45 pm.  Ill plan on finishing up when the tests are all happy (January 21st)

I’ll follow up here once the PR is ready.

**Comment by monigarr** [comment]
All the tests have passed over here. Im getting the PR ready and will keep an eye on this space in case you have any questions for me.

**Comment by monigarr** [comment]
I'm running bugbot checks on my fork for Issue 5066 and will send the PR over as soon as it completes (if it's happy of course).

**Comment by monigarr** [comment]
@fdncred   Thanks for the clarification, it makes sense.

I’m happy to resubmit this work in smaller chunks to make review easier. I’ll split the changes into PRs of ~50 files each, scoped by directory and change type, and keep each PR focused and minimal.

I’ll start with core command directories first and post follow-up PRs incrementally. Each PR will stay strictly within the original Issue #5066 scope (string-only description improvements) and will pass existing tests.

I’ll post the first PR over this weekend :)

**Comment by fdncred** [comment]
Thanks so much @monigarr!! ❤️

**Comment by monigarr** [comment]
I'm working on it today and expect to send smaller PRs of ~50 files each later today.  -edit- Its close to ready as of 4:30 pm e.s.t....

**Comment by monigarr** [comment]
Update Regarding Issue 5066 
in case this is of interest for anyone watching this issue :) 

I split up the previous work for this Issue 5066 into 10 new branches with relevant names each with stricter scope and ~50 file changes max.

I've created each new branch as of 2/3/2026 5pm e.s.t.

I'll be adding each relevant scoped change from my fork branch 'help-text-improvements-5066' as ~50 files for each new branch in the next day or two.

Then I plan to test & verify each of these new branches before sending each new branch as a new PR for issue 5066 over here for review.

\# | Branch name | Scope
-- | ---------------------| -----------------
1 | 5066-help-text-pr-1-nu-cmd-lang | nu-cmd-lang
2 | 5066-help-text-pr-2-nu-cmd-extra | nu-cmd-extra
3 | 5066-help-text-pr-3-nu-command-filters | nu-command filters
4 | 5066-help-text-pr-4-nu-command-formats | nu-command formats
5 | 5066-help-text-pr-5-nu-command-bytes-conversions-db-date | bytes, conversions, database, date, hash, generators, shells
6 | 5066-help-text-pr-6-nu-command-debug-env-exp | debug, env, experimental, misc, help, default_context
7 | 5066-help-text-pr-7-nu-command-fs-path-platform | filesystem, path, platform
8 | 5066-help-text-pr-8-nu-command-math-random | math, random
9 | 5066-help-text-pr-9-nu-command-strings | strings
10 | 5066-help-text-pr-10-nu-command-network-system-viewers | network, system, viewers, nu-command tests

**Comment by fdncred** [comment]
Sounds great, thanks for your help!

**Comment by monigarr** [comment]
I'll plan to send up 3 new PRs for this one Issue 5066 after the previous 3 are reviewed, as time allows until all 10 PRs for this issue are done. 

Maybe that will be a little easier on everyone's work schedules?

**Comment by fdncred** [comment]
ya, thanks. i'm totally swamped right now so i may not get to it for a few days.

**Comment by monigarr** [comment]
@fdncred   I’ll be patient. Thanks for looking at it as you get time..  I really enjoy using nushell and happy to help out as I can too :)

**Comment by monigarr** [comment]
@fdncred   I have 7 more PRs of ~50 files each to send for this issue

**Comment by fdncred** [comment]
I'm glad I was able to look at some today.
